datafusion/core/tests/sqllogictests/src/engines/postgres/mod.rs

```diff
@@ -29,69 +30,157 @@ use postgres_types::Type;
+    ///
+    /// This defaults to
+    ///
+    /// ```text
+    /// PG_DSN="postgresql://postgres@127.0.0.1/test"
+    /// ```
+    ///
+    /// See https://docs.rs/tokio-postgres/latest/tokio_postgres/config/struct.Config.html#url for format
+    pub async fn connect(file_name: impl Into<String>) -> Result<Self> {
+        let file_name = file_name.into();
+
+        let dsn = if let Ok(val) = std::env::var("PG_DSN") {
+            val
+        } else {
+            PG_DSN.to_string()
+        };
+
+        debug!("Using posgres dsn: {dsn}");
+
+        let config = tokio_postgres::Config::from_str(&dsn)?;
+
         let mut retry = 0;
+
         loop {
-            let connection_result =
-                Postgres::connect(file_name.clone(), host, port, db, user, pass).await;
+            let connection_result = config.connect(tokio_postgres::NoTls).await;
+
             match connection_result {
                 Err(e) if retry <= 3 => {
                     debug!("Retrying connection error '{:?}'", e);
                     retry += 1;
                     tokio::time::sleep(Duration::from_secs(1)).await;
                 }
-                result => break result,
+                Err(e) => return Err(Error::from(e)),
+                Ok((client, connection)) => {
+                    let join_handle = tokio::spawn(async move {
+                        if let Err(e) = connection.await {
+                            log::error!("Postgres connection error: {:?}", e);
+                        }
+                    });
+
+                    return Ok(Self {
+                        client,
+                        join_handle,
+                        file_name,
+                    });
+                }
             }
         }
     }
 
-    async fn connect(
-        file_name: String,
-        host: &str,
-        port: u16,
-        db: &str,
-        user: &str,
-        pass: &str,
-    ) -> Result<Self, tokio_postgres::error::Error> {
-        let (client, connection) = tokio_postgres::Config::new()
-            .host(host)
-            .port(port)
-            .dbname(db)
-            .user(user)
-            .password(pass)
-            .connect(tokio_postgres::NoTls)
-            .await?;
-
-        let join_handle = tokio::spawn(async move {
-            if let Err(e) = connection.await {
-                log::error!("Postgres connection error: {:?}", e);
+    /// Special COPY command support. "COPY 'filename'" requires the
+    /// server to read the file which may not be possible (maybe it is
+    /// remote or running in some other docker container).
+    ///
+    /// Thus, we rewrite  sql statements like
+    ///
+    /// ```sql
+    /// COPY ... FROM 'filename' ...
+    /// ```
+    ///
+    /// Into
+    ///
+    /// ```sql
+    /// COPY ... FROM STDIN ...
+    /// ```
+    ///
+    /// And read the file locally.
+    async fn run_copy_command(&mut self, sql: &str) -> Result<DBOutput> {
```

Comment:
I was quite pleased with this :bowtie:  -- it allows the tests to run `COPY FROM 'file'`